streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -50,65 +57,135 @@
+            // completed yet; do not start the first transaction until the topology has been initialized later
+            try {
+                producer.initTransactions();
+            } catch (final TimeoutException exception) {
+                final String errorMessage = "Timeout exception caught when initializing transactions for task " + taskId + ". " +
+                    "\nThe broker is either slow or in bad state (like not having enough replicas) in responding to the request, " +
+                    "or the connection to broker was interrupted sending the request or receiving the response. " +
+                    "\n Consider overwriting `max.block.ms` to a larger value to avoid timeout errors";
 
-        if (partitioner != null) {
-            final List<PartitionInfo> partitions = producer.partitionsFor(topic);
-            if (partitions.size() > 0) {
-                partition = partitioner.partition(topic, key, value, partitions.size());
-            } else {
-                throw new StreamsException("Could not get partition information for topic '" + topic + "'." +
-                    " This can happen if the topic does not exist.");
+                // TODO K9113: we do NOT try to handle timeout exception here but throw it as a fatal error
+                throw new StreamsException(errorMessage, exception);
+            } catch (final KafkaException exception) {
+                throw new StreamsException("Error encountered while initializing transactions for task " + taskId, exception);
+            }
+        }
+    }
+
+    private void maybeBeginTxn() {
+        if (eosEnabled && !transactionInFlight) {
+            try {
+                producer.beginTransaction();
+            } catch (final ProducerFencedException error) {
+                throw new TaskMigratedException(taskId, "Producer get fenced trying to begin a new transaction", error);
+            }
+            transactionInFlight = true;
+        }
+    }
+
+    private void maybeAbortTxn() {
+        if (eosEnabled && transactionInFlight) {
+            try {
+                producer.abortTransaction();
+            } catch (final ProducerFencedException ignore) {
+                /* TODO
+                 * this should actually never happen atm as we guard the call to #abortTransaction
+                 * -> the reason for the guard is a "bug" in the Producer -- it throws IllegalStateException
+                 * instead of ProducerFencedException atm. We can remove the isZombie flag after KAFKA-5604 got
+                 * fixed and fall-back to this catch-and-swallow code
+                 */
+
+                // can be ignored: transaction got already aborted by brokers/transactional-coordinator if this happens
+            }
+            transactionInFlight = false;
+        }
+    }
+
+    public void commit(final Map<TopicPartition, OffsetAndMetadata> offsets) {
```

Comment:
q: I have the same curiosity before, but if the offsets are not moving forward from last iteration, could we skip the commit as an optimization?

Reply:
Yes we could -- my vague memory is that we used to do it on the stream-thread level, but it was later regressed because it is too complicated. I think maybe now is a better time to pick up this optimization again.